### DISCUSSION TITLE
close_session required?

### DISCUSSION THREAD

**Discussion by wollud1969** [FreeOpcUa/opcua-asyncio#820] [Q&A] [answered]
Hi,

I'm using objects of the Client class together with a context manager to talk to an OpenPCS7 server:

```
while not stopped:
    async with Client(url=...) as client:
        for nodeSpec in nodes:
            try:
                node = client.get_node(nodeSpec)
                dataValue = await node.read_data_value()
                # do something use with the dataValue
            except ua.UaError as e:
                # more or less just log the caught exception
    await asyncio.sleep(period)
```

I've started with the "client minimal" example and introduce my own data handling.

My assumption is that using the context manager is enough for tidying up all used resources. However, occasionally - to be honest: quite often, I get `BadTooManySessions` errors from the server. Is it, beyond using the context manager, required to call `activate_session` and particularly `close_session`? Or is that, as I assume, done by the context manager for me? In that case I would reckon that the server is somehow to "small", that there are a couple of other clients and that it is not really my mistake to get this errors.

Thanks and cheers
Wolfgang

**Comment by swamper123** [comment]
First: I am not familiar with OpenPCS7 server.

The context manager calls [__aexit__](https://github.com/FreeOpcUa/opcua-asyncio/blob/560eb91e91efe94f1da46a032e340c81e344b4e3/asyncua/client/client.py#L71) which should [disconnects](https://github.com/FreeOpcUa/opcua-asyncio/blob/560eb91e91efe94f1da46a032e340c81e344b4e3/asyncua/client/client.py#L254) the Client. What I can't see is a 

You may have to take a look what the server limits are...not sure how your setup is built atm.

**Reply by swamper123** [reply]
@wollud1969  Are any Exceptions thrown or errors logged, while closing the connection in normal cases?

**Reply by wollud1969** [reply]
No, there are no other exceptions. Suddenly the BadTooManySessions pops up.

**Comment by oroulet** [comment, marked as answer]
interesting issue. the context manager calls disconnect that calls close_session. So in theory the session is closed...

**Comment by wollud1969** [comment]
So, at least I'm on the right side when using the context manager and not closing the session explicitly. That's fine. Considering some other quite interesting behaviors of the OpenPCS7 server I would assume for the moment that it is not this library or my own code which is doing something wrong.

**Reply by swamper123** [reply]
In case to be 100% sure, you can trace the traffic via Wireshark and see, if everything is done as expected from Client side.
